### PR TITLE
Add JSON logging to graph node

### LINKED ISSUE #3853: How to run graph-node with json logging
**Do you want to request a *feature* or report a *bug*?**

Feature (?)

**What is the current behavior?**

Graph node currently does not do JSON logging and the [environment variables](https://github.com/graphprotocol/graph-node/blob/master/docs/environment-variables.md) don't seem to offer a configuration for that

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem.**

No steps to reproduce, just run the graph node

**What is the expected behavior?**

I expect (?) that there's a configuration option for JSON logging

### PR DESCRIPTION
**Pull request by paymog**
Add JSON logging to graph and make it controllable by an env var, `JSON_LOGGING`.

Fixes #3853

### CODE REVIEW

**Comment by neysofu** [comment]
Hi @paymog! Thanks for the PR.

I'm conflicted on this. On the one hand, I appreciate the fact that JSON logging facilitates integration with some log analytics platforms e.g. Datadog; on the other, this is a niche configuration setting that may result in fragmentation of log formats used in bug reports. Would you mind expanding a bit on your use case, please?

**Comment by paymog** [comment]
> Hi @paymog! Thanks for the PR.
> 
> I'm conflicted on this. On the one hand, I appreciate the fact that JSON logging facilitates integration with some log analytics platforms e.g. Datadog; on the other, this is a niche configuration setting that may result in fragmentation of log formats used in bug reports. Would you mind expanding a bit on your use case, please?

Hey! Yes, this is precisely to make log ingestion/analysis through platforms like datadog easier. 

Do you mind expanding on why this would be a niche configuration? Do most folks who run graph-node not use log ingestion/analytics platforms? Is there specific data about how many (or few) graph node hosts don't use platforms like datadog for monitoring?

**Comment by neysofu** [comment]
> Do you mind expanding on why this would be a niche configuration? Do most folks who run graph-node not use log ingestion/analytics platforms?

 I definitely expect most indexers to use log explorers, but most (all?) of these platforms allow you to parse log messages with custom rules, thus they accept arbitrary log formats. The set of users that *want* to use log analytics and explorer platforms but *can't* because of a lack of JSON logging is, I would imagine, very small.

**Comment by paymog** [comment]
@neysofu ah yes, I totally agree! The set of users that want to use log analytics and explorer platforms but can't because of a lack of JSON logging is probably very small.

I've also set up a custom parser on datadog. However I find that this custom parser takes quite a bit of time to set up and even then isn't perfect (though that may just be because I haven't spent enough time on it) and is also something I may have to maintain over time if the logs change in some way.

As a follow up question, is it efficient as an ecosystem to have each indexer come up with the relevant parsing expression for their log platform of choice? That seems like quite a bit of duplicated effort for each indexer with not much benefit. 

I guess one alternative to this PR would be to keep track of the correct parsing expression for various platforms. That way when a new indexer decides to join, they can go to a well maintained list and more-or-less copy-paste a parsing expression into their preferred logging platform.

However, I still believe json logging will have far superior results because it means those parsing expressions don't need to be maintained and it means that platforms which don't have highly configurable parsers can still be used.

File: graph/src/log/mod.rs [hunk 1]
```diff
@@ -17,25 +28,24 @@ macro_rules! impl_slog_value {
...
-
-use crate::prelude::ENV_VARS;
-
 pub mod codes;
 pub mod elastic;
 pub mod factory;
 pub mod split;
 
-pub fn logger(show_debug: bool) -> Logger {
+pub fn logger(show_debug: bool, json: bool) -> Logger {
```
Comment (mangas) [inline graph/src/log/mod.rs:36]: I think this should take an Option<DrainPicker> if None, default to CustomFormat, if one passed then use that instead

Comment (mangas) [reply graph/src/log/mod.rs:36]: you can add a `json_logger` function which does that setup and then calls `logger(debug, drain)`

File: node/src/opt.rs [hunk 2]
```diff
@@ -163,6 +163,13 @@ pub struct Opt {
     pub expensive_queries_filename: String,
     #[clap(long, help = "Enable debug logging")]
     pub debug: bool,
+    #[clap(
+        long,
+        value_name = "JSON_LOGGING",
+        env = "JSON_LOGGING",
+        help = "output logs in json format"
+    )]
+    pub json_logging: bool,
```
Comment (mangas) [inline node/src/opt.rs:172]: This should also be a configurable option. Could have the values: `default` and `json` for now, others can be added later.

File: graph/src/log/mod.rs (hunk 1 above)
Comment (mangas) [reply graph/src/log/mod.rs:36]: could also setup an helper `default_logger` this way the other instances can just be replaced from `logger` to `default_logger` on tests and such

**Pull request closed without merging**